_posts/2015-03-14-rqlite---building-SQLite-clusters.md

```diff
@@ -0,0 +1,28 @@
+---
+layout: post
+title: rqlite - building SQLite clusters
+date: 2015-03-14 00:00:00
+description:
+author: oscillatingworks
+hn-discussion:
+tags:
+- raft
+- sqlite
+- iot
+- paxos
+- distributed consensus
+---
+
+Today our compass lead us into direction of [rqlite](https://github.com/otoolep/rqlite). It a is small, comprehensive, [Raft](https://raftconsensus.github.io/) based clustering middleware for [SQLite](https://sqlite.org/) databases. Someone could ask who the hell would need SQLite cluster but first things first.
+
+Rqlite is using Raft [consensus protocol](http://en.wikipedia.org/wiki/Consensus_%28computer_science%29) to perform replication among SQLite database nodes. In short Raft is an alternative to [Paxos](http://en.wikipedia.org/wiki/Paxos_%28computer_science%29) which is hardly understood among population including ourselves. In general Raft ensures that if any state machine applies some operation as the `nth command`, no other state machine will ever apply a different `nth command`. This leads to consistency among distributed state machines and this is something Rqlite guaranties among SQLite nodes - consistency and partition tolerance - CP among [CAP](http://de.wikipedia.org/wiki/CAP-Theorem).
+
+Now we know what rqlite is doing and more or less how. But lets answer most important question - who and why would need it? To answer above questions lets first go back to official SQLite manifesto:
+
+***SQLite is a good fit for use in cellphones, set-top boxes, televisions, game consoles, cameras, watches, kitchen appliances, thermostats, automobiles, machine tools, airplanes, remote sensors, drones, medical devices, and robots: the "internet of things"***
+
+Exactly, kitchen appliances, thermostats, remote sensor, drones, robots - Internet Of Thins - is it not what is hot right now and according to many predictions will be next big thing? We bet it will. 
+
+Currently while implementing IoT solutions you usually use some cloud based data feed and sync mechanisms, i.e [firebase](https://www.firebase.com/), [relayr](https://relayr.io/) but sooner or later your sensors will need to share some data/state only among each other without need to synchronise it over the public network which is faster, cheaper and more reliable. 
+
+SQLite due to its simplicity is perfect to be run on almost any type of connected device. Together with rqlite it can give IoT makers solution to sync any type of data locally. Due to this opportunity we strongly believe that rqlite is a very promising, still fresh but active project. Among that it is implemented in [Go](https://golang.org/) another big thingy in programming languages world. It is really hard to find good reason to not contribute to this small but already interesting project. And we plan to do it. Lets see how far we can go. Will keep you updated.
```

Comment:
implemented in Go, (comma) another new big thing in the world of programming languages.